pkg/smokescreen/conntrack/conn_tracker.go

```diff
@@ -15,19 +19,111 @@ type Tracker struct {
+// ConnSuccessRateStats will be recomputed, and a time window to calculate those statistics over.
+// - calculationInterval is how often statistics will be recomputed, in seconds.
+// - calculationWindow is the period that statistics will be calculated over, in seconds.
+// - cleanupInterval is how often expired items (e.g., items older than the calculationWindow) will be deleted from memory, in seconds.
+func StartNewConnSuccessRateTracker(calculationInterval time.Duration, calculationWindow time.Duration, cleanupInterval time.Duration) *ConnSuccessRateTracker {
+	newSuccessTracker := &ConnSuccessRateTracker{
+		ConnAttempts: cache.New(calculationWindow, time.Second*cleanupInterval),
+	}
+	newSuccessTracker.ConnSuccessRateStats.Store(ConnSuccessRateStats{CalculatedAt: time.Now(), ConnSuccessRate: 100, TotalConns: 0})
+
+	go func() {
+		for {
+			var total, succeeded int
+			for _, success := range newSuccessTracker.ConnAttempts.Items() {
+				total++
+				if success.Object.(bool) {
+					succeeded++
+				}
+			}
+			var successRate float64
+			// Avoid divide by zero errors
+			if total == 0 {
+				successRate = float64(100)
+			} else {
+				successRate = (float64(succeeded) / float64(total)) * 100
+			}
+			newSuccessTracker.ConnSuccessRateStats.Store(ConnSuccessRateStats{CalculatedAt: time.Now(), ConnSuccessRate: successRate, TotalConns: total})
+
+			time.Sleep(calculationInterval)
+		}
+	}()
+
+	return newSuccessTracker
+}
+
+func NewTracker(idle time.Duration, statsc statsd.ClientInterface, logger *logrus.Logger, sd atomic.Value, successRateTracker *ConnSuccessRateTracker) *Tracker {
 	return &Tracker{
-		Map:          &sync.Map{},
-		ShuttingDown: sd,
-		Wg:           &sync.WaitGroup{},
-		IdleTimeout:  idle,
-		statsc:       statsc,
+		Map:                &sync.Map{},
+		ShuttingDown:       sd,
+		Wg:                 &sync.WaitGroup{},
+		IdleTimeout:        idle,
+		statsc:             statsc,
+		SuccessRateTracker: successRateTracker,
+	}
+}
+
+// RecordAttempt stores the result of the most recent connection attempt for a destination.
+func (tr *Tracker) RecordAttempt(dest string, success bool) {
+	if tr.SuccessRateTracker == nil {
+		return
+	}
+	tr.SuccessRateTracker.ConnAttempts.Set(normalizeDomainName(dest), success, cache.DefaultExpiration)
+}
+
+func (tr *Tracker) ReportConnectionSuccessRate() *ConnSuccessRateStats {
+	if tr.SuccessRateTracker != nil {
+		stats := tr.SuccessRateTracker.ConnSuccessRateStats.Load().(ConnSuccessRateStats)
+		return &stats
+	}
+	return nil
+}
+
+// Removes the port number if it exists. If the string passed in is an IP address, the IP address is returned.
+// If it is a hostname, we return the eTLD + 1 if we are able to parse it, or the unchanged hostname otherwise.
+func normalizeDomainName(requested_host string) string {
+	// Strip port number
+	d := strings.Split(requested_host, ":")[0]
+	if net.ParseIP(d) != nil {
+		return d
+	}
+	etldPlusOne, err := publicsuffix.EffectiveTLDPlusOne(d)
```

Comment:
I love this. TIL!

Reply:
Me too! One of those magic moments where after a lot of digging you find something that perfectly solves your problem